code/modules/shield_generators/shield_generator.dm

```diff
@@ -48,22 +51,36 @@
 
 
 /obj/machinery/power/shield_generator/Initialize()
-	. = ..()
+	..()
 	connect_to_network()
-
 	mode_list = list()
 	for(var/st in subtypesof(/datum/shield_mode/))
 		var/datum/shield_mode/SM = new st()
 		mode_list.Add(SM)
+	GLOB.moved_event.register(src, src, .proc/update_overmap_shield_list)
+	. = INITIALIZE_HINT_LATELOAD
 
+/obj/machinery/power/shield_generator/LateInitialize()
+	. = ..()
+	update_overmap_shield_list()
 
 /obj/machinery/power/shield_generator/Destroy()
 	shutdown_field()
 	field_segments = null
 	damaged_segments = null
 	mode_list = null
+	GLOB.moved_event.unregister(src, src)
 	. = ..()
-
+	update_overmap_shield_list()
+
+/obj/machinery/power/shield_generator/proc/update_overmap_shield_list()
+	var/obj/effect/overmap/visitable/current_overmap_object = map_sectors["[z]"]
```

Comment:
So this is actually not an accurate way to check which obj we should be throwing a shield over: if we are on a shuttle with a z with another ship, e.g., this will potentially return that ship rather than our shuttle.

There is a helper for this called `check_ownership`  on overmap objects, which returns true if you properly "belong" to it. For an example of how you are meant to use it see `attempt_hook_up_recursive` on ship computers; it may be time to generalize that.